webui/src/pages/repositories/repository/fileRenderers/index.tsx

```diff
@@ -1,96 +1,342 @@
-import React, { FC, useEffect } from "react";
+import React, {FC, useEffect, useState} from "react";
 import remarkGfm from 'remark-gfm'
 import remarkHtml from 'remark-html'
 import ReactMarkdown from 'react-markdown';
 import { IpynbRenderer as NbRenderer } from "react-ipynb-renderer";
 import SyntaxHighlighter from "react-syntax-highlighter";
 import {githubGist as syntaxHighlightStyle} from "react-syntax-highlighter/dist/esm/styles/hljs";
 
-type RequireAtLeastOne<T, Keys extends keyof T = keyof T> = Pick<T, Exclude<keyof T, Keys>> & {
-    [K in Keys]-?: Required<Pick<T, K>> & Partial<Pick<T, Exclude<Keys, K>>>
-}[Keys];
+import {withConnection} from './duckdb';
+import {Table} from "react-bootstrap";
+import moment from "moment";
+import {useAPI} from "../../../../lib/hooks/api";
+import {objects} from "../../../../lib/api";
+import Alert from "react-bootstrap/Alert";
+import {Error, Loading} from "../../../../lib/components/controls";
+import Form from "react-bootstrap/Form";
+import Button from "react-bootstrap/Button";
+import {DatabaseIcon} from "@primer/octicons-react";
+import {humanSize} from "../../../../lib/components/repository/tree";
 
-interface RendererComponentBase {
-    content: string;
-    contentBlob: Blob;
+
+const MAX_FILE_SIZE = 20971520; // 20MiB
+
+interface RendererComponent {
+    repoId: string;
+    refId: string;
+    path: string;
+    fileExtension: string | null;
+    contentType: string | null;
+    sizeBytes: number;
+}
+
+interface RendererComponentWithText extends RendererComponent{
+    text: string;
 }
 
-export type RendererComponent = RequireAtLeastOne<RendererComponentBase, 'content' | 'contentBlob'>;
 
-export const MarkdownRenderer: FC<RendererComponent> = ({ content }) => {
-    if (content) {
+interface RendererComponentWithDataFormat extends RendererComponent{
+    format: string;
+}
+
+interface RendererComponentWithTextCallback extends RendererComponent {
+    onReady: (text: string) => JSX.Element;
+}
+
+export const DATA_FORMATS = ["parquet", "csv", "tsv"]
+
+export const ObjectRenderer: FC<RendererComponent> = (props: RendererComponent) => {
+    const {fileExtension, contentType, sizeBytes} = props;
+    // is this a data file?
+    const dataFormat = (fileExtension) ? DATA_FORMATS.indexOf(fileExtension) : -1
+    if (dataFormat !== -1) {
+        return <DuckDBRenderer {...props} format={DATA_FORMATS[dataFormat]}/>
+    }
+
+    // other types download the whole thing, which we do not want
+    if (sizeBytes > MAX_FILE_SIZE) {
+        return <ObjectTooLarge {...props}/>
+    }
+    // is this markdown?
+    if (contentType === "text/markdown" || fileExtension === 'md') {
```

Comment:
I think this is overly verbose vs. the previous implementation (look up a map of renders). With the map it's also much quicker and easier to understand what's supported or not.

Reply:
Much more readable now IMHO
If we ever need more complex logic for some keys, we can change the values from `FC<RendererComponent>` to a function or HoC